spark2/src/test/java/org/apache/iceberg/spark/source/TestSelect.java

```diff
@@ -0,0 +1,230 @@
+        .master("local[2]")
+        .getOrCreate();
+  }
+
+  @AfterClass
+  public static void stopSpark() {
+    SparkSession currentSpark = spark;
+    spark = null;
+    currentSpark.stop();
+  }
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private String tableLocation = null;
+
+  @Before
+  public void init() throws Exception {
+    File tableDir = temp.newFolder();
+    this.tableLocation = tableDir.toURI().toString();
+
+    table = TABLES.create(SCHEMA, tableLocation);
+
+    List<Record> rows = Lists.newArrayList(
+        new Record(1, "a", 1.0),
+        new Record(2, "b", 2.0),
+        new Record(3, "c", Double.NaN)
+    );
+
+    Dataset<Row> df = spark.createDataFrame(rows, Record.class);
+
+    df.select("id", "data", "doubleVal").write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    table.refresh();
+
+    Dataset<Row> results = spark.read()
+        .format("iceberg")
+        .load(tableLocation);
+    results.createOrReplaceTempView("table");
+
+    scanEventCount = 0;
+    lastScanEvent = null;
+  }
+
+  @Test
+  public void testSelect() {
+    List<Record> expected = ImmutableList.of(
+        new Record(1, "a", 1.0), new Record(2, "b", 2.0), new Record(3, "c", Double.NaN));
+
+    Assert.assertEquals("Should return all expected rows", expected, sql("select * from table"));
+  }
+
+  @Test
+  public void testSelectRewrite() {
+    List<Record> expected = ImmutableList.of(new Record(3, "c", Double.NaN));
+
+    Assert.assertEquals("Should return all expected rows", expected,
+        sql("SELECT * FROM table where doubleVal = double('NaN')"));
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+  }
+
+  @Test
+  public void testProjection() {
+    List<Record> expected = ImmutableList.of(
+        new Record(1, null, null), new Record(2, null, null), new Record(3, null, null));
+
+    Assert.assertEquals("Should return all expected rows", expected, sql("SELECT id FROM table"));
+
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+    Assert.assertEquals("Should not push down a filter", Expressions.alwaysTrue(), lastScanEvent.filter());
+    Assert.assertEquals("Should project only the id column",
+        table.schema().select("id").asStruct(),
+        lastScanEvent.projection().asStruct());
+  }
+
+  @Test
+  public void testExpressionPushdown() {
+    List<Record> expected = ImmutableList.of(new Record(null, "b", null));
+
+    Assert.assertEquals("Should return all expected rows", expected, sql("SELECT data FROM table WHERE id = 2"));
+
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+    Assert.assertEquals("Should project only id and data columns",
+        table.schema().select("id", "data").asStruct(),
+        lastScanEvent.projection().asStruct());
+  }
+
+  private List<Record> sql(String str) {
+    List<Row> rows = spark.sql(str).collectAsList();
```

Comment:
This seems brittle because it uses types to place the results.

Other tests use `StructProjection` and `StructLikeSet` for similar validations. The incoming row is wrapped to be a `StructLike` and added to a `StructLikeSet` based on the expected schema. Then another `StructLikeSet` is created with the expected rows, which are projected using `StructProjection` and the expected schema. That is a cleaner way to do this, I think.

Reply:
Looks like this uses a Java bean record class, so you could also rely on Spark to convert to your record class, and then use a special comparison function to only compare expected columns.

Reply:
Sounds good, I wanted to scope this `Record` class to only be used for this class' use cases but this is definitely not clean. I changed this to use Spark for converting, but encountered a similar issue as described in [this post](https://stackoverflow.com/questions/44886211/spark-default-null-columns-dataset) that when projecting a subset of columns, conversion doesn't work. Since in this class I'm just projecting one column with primitive type, I convert data frame into their specific classes instead. Please let me know if you know better ways of doing this!